### PR TITLE
Mimecast

### PR DESCRIPTION
**Pull request by surajkadam**
Alert Logic AWS Based API Poll (PAWS) Log Collector Library.

### Solution Description
This repository contains the AWS JavaScript Lambda function and CloudFormation Template (CFT) for deploying a log collector in AWS which will poll a 3rd party service API to collect and forward logs to the Alert Logic CloudInsight backend services.

### Acceptance Criteria for Contributors
- [ ] No errors / warnings on build (including linter)
- [ ] 100% automated test coverage
- [ ] Source layout followed from other collectors
- [ ] No unnecessary code (debug logs, redundant comments, unused blocks of code, etc.)
- [ ] Logging of main steps in the collector’s code
- [ ] Logging of any requests that cause the collector to fail / throw an exception, with reason for failure / debugging info
- [ ] API throttling errors are understood and and properly handled
- [ ] Registration/update/deregistration validated
- [ ] Stats and status validated
- [ ] CloudFormation template specific to setting up the collector, with documentation of any template parameters
- [ ] Collector README, including any set up caveats (both when installed in AL account and in customer’s account) – the user/team should be able to successfully set up, update, remove a collector following the README only
- [ ] Demo of set up, update, removal of a collector, with data shown to be correctly parsed in AL console search
- [ ] Links to API documentation
- [ ] At least temporary access to an environment where RCS can validate collector implementation
- [ ] Contact details in case access to this environment is needed in the future

### CODE REVIEW

File: collectors/mimecast/LICENSE [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Copyright 2019 ALERT LOGIC
```
Comment (kkuzmin) [inline collectors/mimecast/LICENSE:1]: please chenge it to 2021

File: collectors/mimecast/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# Mimecast collector
```
Comment (kkuzmin) [inline collectors/mimecast/README.md:1]: Please add a section about how to configure mimecase credentials. And which Mimecast APIs are used.

File: collectors/mimecast/collector.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        super(context, creds, packageJson.version);
+    }
+
+    pawsInitCollectionState(event, callback) {
+        const startTs = process.env.paws_collection_start_ts ?
+            process.env.paws_collection_start_ts :
+            moment().toISOString();
+        const endTs = moment(startTs).add(this.pollInterval, 'seconds').toISOString();
+
+        const applicationNames = JSON.parse(process.env.collector_streams);
```
Comment (kkuzmin) [inline collectors/mimecast/collector.js:38]: Could you please send a separate PR to update collector [template](https://github.com/alertlogic/paws-collector/tree/master/collectors/template) with stream related changes.

File: collectors/mimecast/README.md (hunk 2 above)
Comment (kkuzmin) [reply collectors/mimecast/README.md:1]: Also some links to mimecast docs will be appreciated.

File: collectors/mimecast/collector.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                }
+            }
+        });
+        return callback(null, initialStates, 1);
+    }
+
+    pawsGetRegisterParameters(event, callback) {
+        const regValues = {
+            mimecastApplicationNames: process.env.collector_streams,
+            mimecastApplicationID: process.env.paws_collector_param_string_1,
```
Comment (kkuzmin) [inline collectors/mimecast/collector.js:63]: please do not pass secrets as registration params. we don't want to store them in the backend

Comment (kkuzmin) [reply collectors/mimecast/collector.js:63]: Is it actually a secret?

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: collectors/mimecast/collector.js (hunk 4 above)
Comment (surajkadam) [reply collectors/mimecast/collector.js:63]: For execute mimecast API we want accessKey(clientID), secretKey(paws secret key), appId (application ID) and appKey (application key).

File: collectors/mimecast/LICENSE (hunk 1 above)
Comment (surajkadam) [reply collectors/mimecast/LICENSE]: Done

File: collectors/mimecast/README.md (hunk 2 above)
Comment (surajkadam) [reply collectors/mimecast/README.md:1]: Done

File: collectors/mimecast/collector.js (hunk 3 above)
Comment (surajkadam) [reply collectors/mimecast/collector.js:38]: I'll do

File: collectors/mimecast/utils.js [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            tsPaths = [{ path: ["datetime"] }];
+            break;
+        case Attachment_Protect_Logs:
+        case URL_Protect_Logs:
+            typeIdPaths = [{ path: ["messageId"] }];
+            tsPaths = [{ path: ["date"] }];
+            break;
+        case Malware_Feed:
+            typeIdPaths = [{ path: ["id"] }];
+            tsPaths = [{ path: ["modified"] }];
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:234]: I think we need to use `create` property as timestamp

File: collectors/mimecast/utils.js [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            typeIdPaths = [{ path: ["aCode"] }];
+            tsPaths = [{ path: ["datetime"] }];
+            break;
+        case Attachment_Protect_Logs:
+        case URL_Protect_Logs:
+            typeIdPaths = [{ path: ["messageId"] }];
+            tsPaths = [{ path: ["date"] }];
+            break;
+        case Malware_Feed:
+            typeIdPaths = [{ path: ["id"] }];
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:233]: I think we need to use `type` property here.

File: collectors/mimecast/cfn/README-MIMECAST.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
+# mimecast-collector.template
+
+Alert Logic Mimecast Log Collector CloudFormation template.
+
+ 
+# Before you begin
+
+This procedure requires administrative permissions in AWS and your Alert Logic 
+Cloud Insight account. You also need to download the Alert Logic custom [CFT](https://github.com/alertlogic/paws-collector/blob/master/collectors/mimecast/cfn/mimecast-collector.template).
```
[the remark this answers is not in the document]
Comment (surajkadam) [reply collectors/mimecast/cfn/README-MIMECAST.md:9]: you mean's this https://s3.amazonaws.com/alertlogic-collectors-us-east-1/cfn/mimecast-collector.template

File: collectors/mimecast/utils.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    let typeIdPaths = [];
+    let tsPaths = [];
+
+    switch (stream) {
+        case Siem_Logs:
+            typeIdPaths = [{ path: ["aCode"] }];
+            tsPaths = [{ path: ["datetime"] }];
+            break;
+        case Attachment_Protect_Logs:
+        case URL_Protect_Logs:
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:228]: Do these two have same format?
In tests I see two examples: `CLICK_LOGS_EVENT ` and `LOG_EVENT ` which have different structure.
Could you please provide examples of logs for these two streams?

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: collectors/mimecast/test/mimecast_mock.js [hunk 9]
```diff
@@ -27,29 +27,40 @@ const AIMS_TEST_CREDS = {
...
+    "senderAddress":"senderAddress",
+    "recipientAddress":"recipientAddress",
+    "fileName":"Test CIS Report 2021-05-25 00:00 Test.pdf",
+    "fileType":"application/pdf",
+    "result":"safe",
+    "actionTriggered":"none",
+    "date":"2021-05-25T00:00:09+0000",
+    "details":"Safe Time taken: 0 hrs, 0 min, 4 sec",
+    "route":"inbound",
+    "messageId":"messageId@email.sssss.com",
```
Comment (kkuzmin) [inline collectors/mimecast/test/mimecast_mock.js:39]: if `messageId` is an email then we don't want to use it as message type.

**Review by kkuzmin — CHANGES_REQUESTED** [review]
@surajkadam  could you please add specific examples of `Attachment_Protect_Logs ` and `URL_Protect_Logs ` like you have it here https://github.com/alertlogic/paws-collector/blob/4cb90d5a1b949b65cb9d2747af16e312f5cea1b9/collectors/mimecast/test/mimecast_mock.js#L76

File: collectors/mimecast/test/mimecast_mock.js (hunk 9 above)
Comment (surajkadam) [reply collectors/mimecast/test/mimecast_mock.js:39]: it's looks like eee1234550eedsdedrfff23eee223eeaa@email.domain.com

Comment (surajkadam) [reply collectors/mimecast/test/mimecast_mock.js:39]: Which message type I use

**Comment by surajkadam** [comment]
> @surajkadam could you please add specific examples of `Attachment_Protect_Logs ` and `URL_Protect_Logs ` like you have it here
> 
> https://github.com/alertlogic/paws-collector/blob/4cb90d5a1b949b65cb9d2747af16e312f5cea1b9/collectors/mimecast/test/mimecast_mock.js#L76

Done

File: collectors/mimecast/utils.js [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+            typeIdPaths = [{ path: ["messageId"] }];
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:231]: let's use `definition` as typeId for Attachment logs
and `category` for URL protect logs.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

**Review by kkuzmin — APPROVED** [review]

**Review by OHaganA — APPROVED** [review]

File: collectors/mimecast/utils.js [hunk 11]
```diff
@@ -40,11 +40,14 @@ function getAPILogs(authDetails, state, accumulator, maxPagesPerInvocation) {
                     pageCount++;
                     switch (state.stream) {
                         case Siem_Logs:
-                            if (body.data) {
+                            if (body.data && body.data[0]) {
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:43]: is it possible to have `body.data` as an empty array? If yes then we'll have an error addressing the first element.

File: collectors/mimecast/utils.js [hunk 12]
```diff
@@ -40,11 +40,14 @@ function getAPILogs(authDetails, state, accumulator, maxPagesPerInvocation) {
...
                                 accumulator.push(...body.data);
                             }
                             if (response.meta && response.meta.isLastToken) {
                                 nextPage = undefined;
+                                if(applicationDetails.payload.data[0].token){
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:48]: same here. Is it possible `applicationDetails.payload.data` is an empty array?

File: collectors/mimecast/utils.js (hunk 11 above)
Comment (surajkadam) [reply collectors/mimecast/utils.js:43]: if isLastToken true then it's possible, And I alreday added testcases for same and it's working

File: collectors/mimecast/utils.js (hunk 12 above)
Comment (surajkadam) [reply collectors/mimecast/utils.js:48]: here, it's not possible because we are generating payload while run API

File: collectors/mimecast/utils.js (hunk 11 above)
Comment (surajkadam) [reply collectors/mimecast/utils.js]: Done

File: collectors/mimecast/utils.js [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+                        return reject(error);
+                    }
+                    body = JSON.parse(body);
+                    if (body.fail && body.fail[0] && body.fail[0].errors) {
+                        return reject(body.fail[0].errors[0]);
+                    }
+                    pageCount++;
+                    switch (state.stream) {
+                        case Siem_Logs:
+                            if (body.data.length > 0) {
```
Comment (kkuzmin) [inline collectors/mimecast/utils.js:43]: I think we need to leave `body.data` check, I mean have
```
if (body.data && body.data.length > 0)
```

**Review by kkuzmin — APPROVED** [review]

File: collectors/mimecast/collector.js [hunk 14]
```diff
@@ -126,14 +127,26 @@ class MimecastCollector extends PawsCollector {
...
+                }
+                
             });
     }
 
     _getNextCollectionState(curState) {
         if (curState.stream === Siem_Logs || curState.stream === Malware_Feed) {
             return {
                 stream: curState.stream,
+                currentMoment: moment().toISOString(),
```
Comment (kkuzmin) [inline collectors/mimecast/collector.js:149]: let's name it like `currentTs`

File: collectors/mimecast/collector.js [hunk 15]
```diff
@@ -126,14 +127,26 @@ class MimecastCollector extends PawsCollector {
                 if (error.code) {
                     error.errorCode = error.code;
                 }
-                return callback(error);
+                if (error.statusCode && error.statusCode == 429) {
+                    error.errorCode = error.statusCode;
```
Comment (kkuzmin) [inline collectors/mimecast/collector.js:131]: not sure why we need this assignment because we don't pass this `error` object anywhere else.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: collectors/mimecast/collector.js [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+                console.info(`MIME000003 Next collection in ${newState.poll_interval_sec} seconds`);
+                return callback(null, accumulator, newState, newState.poll_interval_sec);
+            }).catch((error) => {
+                // set errorCode if not available in error object to showcase client error on DDMetrics
+                if (error.code) {
+                    error.errorCode = error.code;
+                }
+                if (error.statusCode && error.statusCode == 429) {
+                    state.poll_interval_sec = 900;
+                    console.log("The Mimecast service you're trying to access is temporarily busy. Please try again in a few minutes and then contact your IT helpdesk if you still have problems.");
```
Comment (kkuzmin) [inline collectors/mimecast/collector.js:132]: please use `console.warn` here

**Review by kkuzmin — APPROVED** [review]
